cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterAuditLookUp.java

```diff
@@ -0,0 +1,157 @@
+      AuditLogContent destination = entry.getDestination();
+      if (destination.getHttpMethod().equals(httpMethod)) {
+        return destination;
+      }
+    }
+    return null;
+  }
+
+  private void createMatcher() {
+    List<Class<?>> handlerClasses;
+    try {
+      handlerClasses = getAllHandlerClasses();
+    } catch (IOException e) {
+      LOG.error("Failed to get all handler classes for audit logging: {}", e.getCause());
+      return;
+    }
+
+    for (Class<?> handlerClass : handlerClasses) {
+      Path classPath = handlerClass.getAnnotation(Path.class);
+      String classPathStr = classPath == null ? "" : classPath.value();
+      for (Method method : handlerClass.getMethods()) {
+        Path methodPath = method.getAnnotation(Path.class);
+        AuditPolicy auditPolicy = method.getAnnotation(AuditPolicy.class);
+        HttpMethod httpMethod = getHttpMethod(method);
+        if (methodPath == null || auditPolicy == null || httpMethod == null) {
+          continue;
+        }
+
+        String methodPathStr = methodPath.value();
+        String completePath = classPathStr.endsWith("/") || methodPathStr.startsWith("/")
+          ? classPathStr + methodPathStr : classPathStr + "/" + methodPathStr;
+        List<AuditEnum> auditContents = Arrays.asList(auditPolicy.value());
+        List<String> headerNames = new ArrayList<>();
+        if (auditContents.contains(AuditEnum.HEADER)) {
+          Annotation[][] annotations = method.getParameterAnnotations();
+          for (Annotation[] annotationArr : annotations) {
+            if (annotationArr.length > 0) {
+              for (Annotation annotation : annotationArr) {
+                if (annotation instanceof HeaderParam) {
+                  headerNames.add(((HeaderParam) annotation).value());
+                }
+              }
+            }
+          }
+        }
+
+        AuditLogContent auditLogContent = new AuditLogContent(httpMethod,
+                                                              auditContents.contains(AuditEnum.REQUEST_BODY),
+                                                              auditContents.contains(AuditEnum.RESPONSE_BODY),
+                                                              headerNames);
+        patternPathRouterWithGroups.add(completePath, auditLogContent);
+      }
+    }
+  }
+
+  private HttpMethod getHttpMethod(Method method) {
+    if (method.isAnnotationPresent(GET.class)) {
+      return HttpMethod.GET;
+    }
+    if (method.isAnnotationPresent(PUT.class)) {
+      return HttpMethod.PUT;
+    }
+    if (method.isAnnotationPresent(POST.class)) {
+      return HttpMethod.POST;
+    }
+    if (method.isAnnotationPresent(DELETE.class)) {
+      return HttpMethod.DELETE;
+    }
+    return null;
+  }
+
+  private List<Class<?>> getAllHandlerClasses() throws IOException {
+    ClassLoader cl = getClass().getClassLoader();
```

Comment:
For cluster, I think we need to have a integration test to verify that we get all the REST endpoints for audit logging.